### DISCUSSION TITLE
Dynamic tiling manager with Hammerspoon

### DISCUSSION THREAD

**Discussion by AlgusDark** [Hammerspoon/hammerspoon#3254] [Q&A]
I'm wondering if I could ditch tiling managers like Amethyst/Yabai with Hammerspoon. Something I would like to have is the Dynamic tiling when a new window enters/leaves. I'm not sure if there is a watcher for that. Should I use `hs.window.filter.new(true)` and have a global filter and then check all the visible windows?

The idea of my tiling manager would be:
- [ ] Arrange entering/leaving windows as: 
  - [x] Full -- half/half -- half/2-horizontal-split -- 1/4 grid
- [x] Be able to jump from focused window to any other window through cardinal points
- [x] Be able to swap focused window to any cardinal point
- [ ] Be able to create/destroy spaces and move windows through spaces (I believe this is experimental in Hammerspoon... just moving windows to spaces left or right is ok)
- [ ] Open a Space-Domain (something like opening 4 different apps in the current space or a new one)

From the list above, those that aren't ticked are those that I don't know if it's possible to do with hammerspoon 🤔 If is possible, or if there is a spoon for my needs, could anyone share some tips for achieving this?

Thanks!

**Comment by MuhammedZakir** [comment]
See `hs.spaces`[1]. You will definitely have visual-flicker when using this. If you're satisfied with it, fine, otherwise a better approach is to use Yabai, but control it using Hammerspoon - use key mapping modules instead of `skhd` and communicate directly using `hs.socket`[2] rather than by executing `yabai`.

[1] http://www.hammerspoon.org/docs/hs.spaces.html
[2] http://www.hammerspoon.org/docs/hs.socket.html

**Reply by AlgusDark** [reply]
I'm a little confused on how to use `hs.socket` for yabai. Let's say that we have this `yabai -m window --focus west` and I want to bind it to `alt+cmd+h`. How I could do that?

**Comment by MuhammedZakir** [comment]
1. For binding to a keyboard/mouse shortuct, you can use `hs.hotkey`(easy), `hs.hotkey.modal`(easy), or `hs.eventtap`(more work, but is more configurable/customizable).

2. Translating `yabai -m <message>`: using your example, `yabai -m window --focus west`, the message to send will be `window\0--focus\0west\0`.
    - You can construct this manually, or use `string.gsub` (if you don't use space other than as a seperator), or use a table and join them using `table.concat`. See below example (below is just a demonstration, for the last two, you should extract that to a function instead):
    - ```lua
      -- constructing manually
      local message = "window\0--focus\0west\0"
      -- using `string.gsub`
      local message = ("window --focus west "):gsub(" +", "\0")
      -- using `table.concat`
      local message = table.concat({"'window", "--focus", "west"}, "\0") .. "\0"
      ```
3. Because of the way Yabai currently handles socket connection, you will have close the connection each time you write the message and read the response, that means that you will have connect to the socket again before writing message.

4. Be sure to check the first char of the response. `\07` as first char means a failure.

Edit: I didn't post code because I haven't used `hs.socket` before.

Edit-2: Forgot to mention: you will also need to parse successful responses, which will be in JSON. I haven't checked how to do this; you can search for a Lua JSON parser, or you can use LPeg - which will probably enough especially if you only want to cherry-pick some details from the response.

Edit-3: There's already a HS module that provides JSON to Lua-table decoder -- http://www.hammerspoon.org/docs/hs.json.html.

---

Here is an example which uses Go: https://github.com/koekeishiya/yabai/discussions/1312#discussioncomment-3069803. Note that, it uses `\x00` rather than `\0`. Both are same. if you would like to use that, then for `\07`, it would be `\x07` (this is what Yabai uses).

**Reply by MuhammedZakir** [reply]
Converting `yabai -m <commands>`: `client_send_message` function in `yabai.c`: https://github.com/koekeishiya/yabai/blob/32ce2f9590151b37c288cfe5559375bd0e1522e6/src/yabai.c#L46

All possible errors: search for `daemon_fail` in `message.c`: https://github.com/koekeishiya/yabai/blob/master/src/message.c

**Reply by AlgusDark** [reply]
Thanks, I'll try this on the weekend :D

**Reply by brandonkboswell** [reply]
Did any of you try invoking Yabai via socket? Was it significantly different than os.execute?

**Reply by AlgusDark** [reply]
Sorry, I'm now just using yabai but I do use hammerspoon for some tricks.
I showed in this presentation: https://youtu.be/UchF6s2j5r8?t=2269

**Comment by brandonkboswell** [comment]
Thank you sharing Carlos!

**Comment by Bellavene** [comment]
https://github.com/mogenson/PaperWM.spoon
